openrtb_ext/request_test.go

```diff
@@ -141,3 +142,580 @@ var validGranularityTests []granularityTestData = []granularityTestData{
+						Config:  &Config{&ORTB2{Site: map[string]json.RawMessage{"test": json.RawMessage(`{}`)}}},
+					},
+					{
+						Bidders: []string{"Bidder5", "bidder17"},
+						Config:  &Config{&ORTB2{Site: map[string]json.RawMessage{"test": json.RawMessage(`{"foo":"bar"}`), "more": json.RawMessage(`{}`)}}},
+					},
+					{
+						Bidders: []string{"foo"},
+						Config:  &Config{&ORTB2{User: map[string]json.RawMessage{"abc": json.RawMessage(`123`)}}},
+					},
+				},
+			},
+			prebidCopy: &ExtRequestPrebid{
+				BidderConfigs: []BidderConfig{
+					{
+						Bidders: []string{"Bidder1", "bidder2"},
+						Config:  &Config{&ORTB2{Site: map[string]json.RawMessage{"test": json.RawMessage(`{}`)}}},
+					},
+					{
+						Bidders: []string{"Bidder5", "bidder17"},
+						Config:  &Config{&ORTB2{Site: map[string]json.RawMessage{"test": json.RawMessage(`{"foo":"bar"}`), "more": json.RawMessage(`{}`)}}},
+					},
+					{
+						Bidders: []string{"foo"},
+						Config:  &Config{&ORTB2{User: map[string]json.RawMessage{"abc": json.RawMessage(`123`)}}},
+					},
+				},
+			},
+			mutator: func(t *testing.T, prebid *ExtRequestPrebid) {
+				prebid.BidderConfigs[0].Bidders = append(prebid.BidderConfigs[0].Bidders, "bidder4")
+				prebid.BidderConfigs[1] = BidderConfig{
+					Bidders: []string{"george"},
+					Config:  &Config{nil},
+				}
+				prebid.BidderConfigs[2].Config.ORTB2.User["abc"] = json.RawMessage(`{"id": 345}`)
+				prebid.BidderConfigs = append(prebid.BidderConfigs, BidderConfig{
+					Bidders: []string{"bidder2"},
+					Config:  &Config{&ORTB2{}},
+				})
+			},
+		},
+		{
+			name: "Cache",
+			prebid: &ExtRequestPrebid{
+				Cache: &ExtRequestPrebidCache{
+					Bids: &ExtRequestPrebidCacheBids{
+						ReturnCreative: ptrutil.ToPtr(true),
+					},
+				},
+			},
+			prebidCopy: &ExtRequestPrebid{
+				Cache: &ExtRequestPrebidCache{
+					Bids: &ExtRequestPrebidCacheBids{
+						ReturnCreative: ptrutil.ToPtr(true),
+					},
+				},
+			},
+			mutator: func(t *testing.T, prebid *ExtRequestPrebid) {
+				prebid.Cache.Bids = nil
+				prebid.Cache.VastXML = &ExtRequestPrebidCacheVAST{
+					ReturnCreative: ptrutil.ToPtr(true),
+				}
+			},
+		},
+		{
+			name: "Currency",
+			prebid: &ExtRequestPrebid{
+				CurrencyConversions: &ExtRequestCurrency{
+					ConversionRates: map[string]map[string]float64{"A": {"X": 5.4}},
+					UsePBSRates:     ptrutil.ToPtr(false),
+				},
+			},
+			prebidCopy: &ExtRequestPrebid{
+				CurrencyConversions: &ExtRequestCurrency{
+					ConversionRates: map[string]map[string]float64{"A": {"X": 5.4}},
+					UsePBSRates:     ptrutil.ToPtr(false),
+				},
+			},
+			mutator: func(t *testing.T, prebid *ExtRequestPrebid) {
+				prebid.CurrencyConversions.ConversionRates["A"]["X"] = 3.4
+				prebid.CurrencyConversions.ConversionRates["B"] = make(map[string]float64)
+				prebid.CurrencyConversions.ConversionRates["B"]["Y"] = 0.76
+				prebid.CurrencyConversions.UsePBSRates = ptrutil.ToPtr(true)
+			},
+		},
+		{
```

Comment:
Can we please also add a test case for `ExtRequestPrebid.Data`?

Reply:
The general test has the case of a nil data being populated, but I can do the other case as well.